src/shacl/ShaclParser.ts

```diff
@@ -0,0 +1,583 @@
+    cst: any;
+  } => {
+    this.input = this.tokenize(document);
+    const cst = this.turtleDoc();
+    const { validationErrors } = this.validateWithVisitor(cst);
+    // Next two items are copied so that they can be returned/held after parse
+    // state is cleared.
+    const errors: IRecognitionException[] = [...this.errors];
+    const semanticErrors: IRecognitionException[] = [
+      ...this.semanticErrors,
+      ...validationErrors,
+    ];
+
+    return {
+      errors,
+      semanticErrors,
+      cst,
+    };
+  };
+
+  constructor(
+    config?: Partial<IParserConfig>,
+    prefixes = { shacl: 'sh', xsd: 'xsd' }
+  ) {
+    super(
+      {
+        outputCst: true,
+        recoveryEnabled: true,
+        ...config,
+      },
+      getShaclTokenTypes(prefixes),
+      getShaclTokenTypes(prefixes),
+      false
+    );
+
+    this.prefixes = prefixes;
+    this.lexer = new Lexer(getShaclTokenTypes(prefixes));
+    this.shaclTokenMap = getShaclTokenMap(prefixes);
+    Parser.performSelfAnalysis(this);
+  }
+
+  predicateObjectList = this.OVERRIDE_RULE('predicateObjectList', () => {
+    this.OR([
+      {
+        ALT: () => {
+          this.SUBRULE(this.shaclRulePredicateObjectList);
+        },
+      },
+      {
+        ALT: () => {
+          this.SUBRULE(this.verb);
+          this.SUBRULE(this.objectList);
+        },
+      },
+    ]);
+    this.MANY(() => {
+      this.CONSUME(turtleTokenMap.Semicolon);
+      this.OPTION(() => {
+        this.OR1([
+          {
+            ALT: () => {
+              this.SUBRULE1(this.shaclRulePredicateObjectList);
+            },
+          },
+          {
+            ALT: () => {
+              this.SUBRULE1(this.verb);
+              this.SUBRULE1(this.objectList);
+            },
+          },
+        ]);
+      });
+    });
+  });
+
+  shaclRulePredicateObjectList = this.RULE(
+    'shaclRulePredicateObjectList',
+    () => {
+      this.OR([
+        {
+          ALT: () => this.SUBRULE(this.shaclPredicateIRI),
+        },
+        {
+          ALT: () => this.SUBRULE(this.shaclNodeKind),
+        },
+        {
+          ALT: () => this.SUBRULE(this.shaclTargetNode),
+        },
+        {
+          ALT: () => this.SUBRULE(this.shaclPropertyPath),
+        },
+        {
+          ALT: () => this.SUBRULE(this.shaclLiteralConstraint),
+        },
+        {
+          ALT: () => this.SUBRULE(this.shaclListTakingConstraint),
+        },
+        {
+          ALT: () => this.SUBRULE(this.shaclShapeExpectingConstraint),
+        },
+        {
+          ALT: () => this.SUBRULE(this.shaclHasValueConstraint),
+        },
+        {
+          ALT: () => this.SUBRULE(this.shaclVerbShape),
+        },
+      ]);
+    }
+  );
+
+  shaclPredicateIRI = this.RULE('shaclPredicateIRI', () => {
+    this.CONSUME(categoryTokenMap.IriTakingPredicate);
+    this.SUBRULE(this.iri); // FIXME? This may need to allow a list of IRIs, as with shaclTargetNode.
```

Comment:
Am I looking at the right thing for this? https://www.w3.org/TR/shacl/#property-shapes
> A shape has at most one value for sh:path. Each value of sh:path in a shape must be a well-formed SHACL property path.
which says the answer is don't fixme, this is right, it's a single IRI or a property path, no comma-separated list

while `sh:targetNode` definitely accepts a command-separated list of IRIs

Reply:
BTW - links to the spec for each of these rules would be soooo helpful for trying to learn what's going on. Though I do realize the names mostly almost match (when you namespace and change casing)

Reply:
The point of the comment was that we _may_ need to do the same thing as we do in the `shaclTargetNode` rule, which accepts a list of IRIs.

The rule you're commenting on applies to these tokens:

```
  IriTakingPredicate: getAsTypedTuple(
    'class',
    'datatype',
    'equals',
    'disjoint',
    'lessThan',
    'lessThanOrEquals',
    'targetClass',
    'targetSubjectsOf',
    'targetObjectsOf',
    'severity'
  ),
```

I think none of them accepts a list, though there are weird cases where the SHACL tests (though maybe not the spec) accept single-element collections (e.g., `sh:severity (sh:Info)`). I think I should probably just delete the comment, though. Do you see an issue here?

Reply:
Also, linking to the rule is hard because of the number of predicates covered by various rules. SHACL doesn't provide a formalized EBNF for each serialization, so there isn't really a rule in the grammar corresponding just to 'shaclIRIPredicate'.

Reply:
Ah, I see.

For `sh:class` the spec does have this note:
> Note that multiple values for sh:class are interpreted as a conjunction, i.e. the values need to be SHACL instances of all of them. 

`sh:datatype` is explicitly only 1

`sh:targetClass` says this which implies there can be multiple
>  Each value of sh:targetClass in a shape is an IRI.

the others aren't very clear

Reply:
Yeah, but when the spec says that they can have multiple values, another possibility is just this:

```
:SomeThing a sh:NodeShape ;
  sh:targetClass :SomeClass ;
  sh:targetClass :SomeOtherClass .
```

The value itself in this case is always just "an IRI."

Even so, I'm becoming inclined to think that the comma syntax is purely shorthand for that and should be allowed, too. So maybe I'll revisit this rule.

Reply:
right, that's my understanding of the comma syntax

Reply:
https://www.w3.org/TR/turtle/#object-lists